SourceX/miniwin/misc_msg.cpp

```diff
@@ -333,6 +344,9 @@ void StoreSpellCoords()
 			xo = END_X;
 			yo -= BOX_SIZE;
 		}
+#ifndef USE_SDL1
+		finish_simulated_mouse_clicks(MouseX, MouseY);

```

Comment:
A `for` loop body in an unrelated function doesn't seem like the right place for this

Reply:
Somebody messed this up, this was originally in a different place
```
WINBOOL PeekMessageA(LPMSG lpMsg, HWND hWnd, UINT wMsgFilterMin, UINT wMsgFilterMax, UINT wRemoveMsg)
{
	// update joystick mouse at maximally 60 fps
	// update joystick (and touch mouse on Switch) at maximally 60 fps
	currentTime = SDL_GetTicks();
	if ((currentTime - lastTime) > 15) {
#ifdef SWITCH
		switch_finish_simulated_mouse_clicks(MouseX, MouseY);
#endif
		HandleJoystickAxes();
		lastTime = currentTime;
	}```

Who is messing with the Switch code so much, with seemingly no understanding how it works?

Reply:
It would be nice if people who mess with Switch code actually would test their changes. There are several Switch emulators out there that can be used for testing.

Reply:
> Who is messing with the Switch code so much, with seemingly no understanding how it works?

I don't think anybody's messing with it, just rebasing correctly is hard because there were so many changes to upstream since the Switch port. Please review vigilantly :)

Aside: When I tried to use the emulator it didn't work correctly, though users reported it working on a real device. Which emulator/version do you use to test?

Reply:
Once per frame, switch_finish_simulated_mouse_clicks(MouseX, MouseY) must be called with the current mouse coordinates. On Switch, current mouse coordinates are not the same as SDL Mouse coords if I remember right, because the SDL Mouse pointer stuff doesn't work.

Reply:
> Who is messing with the Switch code so much, with seemingly no understanding how it works?

I am, the Switch code broke several other platforms, messed with code it shouldn't and did not integrate things well. Sorting all this out is a lot of work, and hard to do without access to hardware etc. But not one else appears to be willing to do the needed work. Occasionally I make mistake, but that is why I create separate branches so that I can go over it and have people review and test for me. It's been less than 3 hours since I published this latest rebase so I haven't had much chance to go over things. It should be fixed now and in a better way than was done originally. Please test.

Reply:
I used Yuzu before, but that emulator only works on Windows with decent GL support. It could very well be that Switch functionality was broken recently, and/or that people who tested on real hardware didn't bother to test touch gestures (the game can be played without touch, but it is fun to be able to use touch).

Reply:
I remember I tested not too long ago, on October 7, when AJenbo pinged me about the cmake build. Back then it all worked and I know I tested touch controls.

Reply:
> the SDL Mouse pointer stuff doesn't work. I am going by memory here since I haven't looked at this code in a while.

This was broken early on in the Switch port (https://github.com/lantus/devilution-nx/commit/a03fa8f84f6490d594eb45d782b4e2530fb8cd9d#diff-6607e4567dd54a4cab274488d43bed4eR10) and several hacks where applied on top instead of solving it. It would be really helpful if you can test it out to see if it will work correctly now that all the other code has been cleaned up.

Reply:
I just tested latest build, indeed simulated (tap) mouse clicks are never "lifted" so this is currently broken.

Reply:
> I remember I tested not too long ago, on October 7, when AJenbo pinged me about the cmake build. Back then it all worked and I know I tested touch controls.

@rsn8887 October 7 was the previous time that I did a rebase of the switch code, it was still not ready to be merged at that point (though much better than previous). So no one has touched it since then, today I am rebasing it again to bring it up to date and merge it with the new controller code.

Unfortunately I don't have much time to setup and emulator etc, so you being able to test it when ever I do this work is extreamly helpful to the process.